apps/spark/src/main/java/com/linkedin/openhouse/jobs/scheduler/tasks/OperationTasksBuilder.java

```diff
@@ -19,46 +20,47 @@
 @AllArgsConstructor
 public class OperationTasksBuilder {
   @Getter(AccessLevel.NONE)
-  private final OperationTaskFactory<? extends OperationTask> taskFactory;
+  private final OperationTaskFactory<? extends OperationTask<?>> taskFactory;
 
   protected final TablesClient tablesClient;
 
-  private List<OperationTask> prepareTableOperationTaskList(JobConf.JobTypeEnum jobType) {
-    List<OperationTask> taskList = new ArrayList<>();
-    for (TableMetadata metadata : tablesClient.getTables()) {
-      log.info("metadata: {}", metadata);
-      try {
-        taskList.add(taskFactory.create(metadata));
-      } catch (Exception e) {
-        throw new RuntimeException(
-            String.format(
-                "Cannot create operation task from TableMetadata: %s given job type: %s",
-                metadata, jobType),
-            e);
-      }
-    }
-    return taskList;
+  private List<OperationTask<?>> prepareTableOperationTaskList(JobConf.JobTypeEnum jobType) {
+    List<TableMetadata> tableMetadataList = tablesClient.getTableMetadataList();
+    log.info("Fetched metadata for {} tables", tableMetadataList.size());
+    return processMetadataList(tableMetadataList, jobType);
+  }
+
+  private List<OperationTask<?>> prepareTableDirectoryOperationTaskList(
+      JobConf.JobTypeEnum jobType) {
+    List<DirectoryMetadata> directoryMetadataList = tablesClient.getOrphanTableDirectories();
+    log.info("Fetched metadata for {} directories", directoryMetadataList.size());
+    return processMetadataList(directoryMetadataList, jobType);
   }
 
-  private List<OperationTask> prepareTableDirectoryOperationTaskList(JobConf.JobTypeEnum jobType) {
-    List<OperationTask> taskList = new ArrayList<>();
-    for (DirectoryMetadata metadata : tablesClient.getOrphanTableDirectories()) {
-      log.info("metadata: {}", metadata);
+  private List<OperationTask<?>> processMetadataList(
+      List<? extends Metadata> metadataList, JobConf.JobTypeEnum jobType) {
+    List<OperationTask<?>> taskList = new ArrayList<>();
+    for (Metadata metadata : metadataList) {
+      log.info("Found metadata {}", metadata);
       try {
-        taskList.add(taskFactory.create(metadata));
+        OperationTask<?> task = taskFactory.create(metadata);
+        if (!task.shouldRun()) {
+          log.info("Skipping task {}", task);
```

Comment:
We are relying on skipping job in OperationTask for metrics. This looks like we are introducing 2 control paths for skipping?

Reply:
It's still in one place and moved to here because the check has become very cheap, so it doesn't need to run on a thread-pool. The subsequence handling of counts in jobs scheduler will become cleaner.

Reply:
I'll do Job scheduler clean up in the subsequent PR

Reply:
So is it right to say Total tables = Number of jobs created + skipped ones  (from this log) ?

Reply:
Sure. database filter should be fine